File: rust/day8/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader, Result as IOResult};

const INPUT_PATH: &str = "../inputs/day8.txt";

type Tree = u32;
type Forest = Vec<Vec<Tree>>;
type ScenicScore = usize;

fn read_forest() -> IOResult<Forest> {
    Ok(BufReader::new(File::open(INPUT_PATH)?)
        .lines()
        .flat_map(|line| line)
        .map(|line| {
            line.chars()
                .map(|t| t.to_digit(10).expect("Line only contains numbers"))
                .collect()
        })
        .collect())
}

fn is_visible(forest: &Forest, (row, col): (usize, usize)) -> bool {
    let (height, width) = (forest.len(), forest[0].len());

    let from_left = (0..col).all(|c| forest[row][c] < forest[row][col]);
    let from_right = (col + 1..width).all(|c| forest[row][c] < forest[row][col]);
    let from_top = (0..row).all(|r| forest[r][col] < forest[row][col]);
    let from_bottom = (row + 1..height).all(|r| forest[r][col] < forest[row][col]);

    from_left || from_right || from_top || from_bottom
}

fn scenic_score(forest: &Forest, (row, col): (usize, usize)) -> ScenicScore {
    let (height, width) = (forest.len(), forest[0].len());

    let from_left = col
        - (0..col)
            .rev()
            .find(|c| forest[row][*c] >= forest[row][col])
            .unwrap_or(0);
    let from_right = (col + 1..width)
        .find(|c| forest[row][*c] >= forest[row][col])
        .unwrap_or(width - 1)
        - col;
    let from_top = row
        - (0..row)
            .rev()
            .find(|r| forest[*r][col] >= forest[row][col])
            .unwrap_or(0);
    let from_bottom = (row + 1..height)
        .find(|r| forest[*r][col] >= forest[row][col])
        .unwrap_or(height - 1)
        - row;

    from_left * from_right * from_top * from_bottom
}

fn count_visible_trees(forest: &Forest) -> usize {
    let (height, width) = (forest.len(), forest[0].len());
    (0..height).fold(0, |row_acc, row| {
        row_acc
            + (0..width).fold(0, |col_acc, col| {
                col_acc + is_visible(&forest, (row, col)) as usize
            })
    })
}

fn max_scenic_score(forest: &Forest) -> ScenicScore {
    let (height, width) = (forest.len(), forest[0].len());
    (0..height)
        .reduce(|row_max, row| {
            (0..width)
                .reduce(|col_max, col| scenic_score(forest, (row, col)).max(col_max))
                .expect("Column has a maximal scenic score")
                .max(row_max)
        })
        .expect("Forest has a maximal scenic score")
}

fn main() -> IOResult<()> {
    let forest = read_forest()?;
    assert!(
        forest.iter().all(|row| row.len() == forest[0].len()),
        "Forest rows are not of the same size"
    );

    println!("Task 1: {}", count_visible_trees(&forest));
    println!("Task 2: {}", max_scenic_score(&forest));
    Ok(())
}

File: ocaml/day8/main.ml
open Core

let ( << ) = Fn.compose
let ( --> ) = List.range
let ( <-- ) a b = List.rev (a --> b)
let max l ~compare = Option.value_exn @@ List.max_elt ~compare l

let trees =
  In_channel.read_lines "../inputs/day8.txt"
  |> Array.of_list_map ~f:(Array.map ~f:Char.get_digit_exn << String.to_array)

let h, w = (Array.length trees, Array.length trees.(0))

let is_visible r c =
  let f c' = trees.(r).(c') < trees.(r).(c) in
  let horizontally = List.for_all (0 --> c) ~f || List.for_all (c + 1 --> w) ~f in
  let f r' = trees.(r').(c) < trees.(r).(c) in
  let vertically = List.for_all (0 --> r) ~f || List.for_all (r + 1 --> h) ~f in
  horizontally || vertically

let scenic_score r c =
  let f c' = trees.(r).(c') >= trees.(r).(c) in
  let from_left = Option.value (List.find ~f (0 <-- c)) ~default:0 in
  let from_right = Option.value (List.find ~f (c + 1 --> w)) ~default:(w - 1) in
  let f r' = trees.(r').(c) >= trees.(r).(c) in
  let from_above = Option.value (List.find ~f (0 <-- r)) ~default:0 in
  let from_below = Option.value (List.find ~f (r + 1 --> h)) ~default:(h - 1) in
  (c - from_left) * (from_right - c) * (r - from_above) * (from_below - r)

let num_visible_trees =
  let f r = List.count (0 --> w) ~f:(is_visible r) in
  List.sum (module Int) (0 --> h) ~f

let max_scenic_score =
  let f r = max (List.map (0 --> w) ~f:(scenic_score r)) ~compare:Int.compare in
  max (List.map (0 --> h) ~f) ~compare:Int.compare

let () =
  printf "Task 1: %d\n" @@ num_visible_trees;
  printf "Task 2: %d\n" @@ max_scenic_score
